### PR TITLE
MLM adaptation and Multitask Finetuning

### PR DESCRIPTION
**Pull request by lintangsutawika**

### CODE REVIEW

File: 4B8-en-ND-MLM.sh [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+NNODES=1
+PP_SIZE=2 # NLAYERS must be a multiple of PP_SIZE here
+TP_SIZE=1 # always fixed to the size of a single node
+DP_SIZE=$((NNODES*GPUS_PER_NODE/(PP_SIZE*TP_SIZE))) # will get derived automatically by trainer
+
+MICRO_BATCH_SIZE=1
+GLOBAL_BATCH_SIZE=512
+TRAIN_ITER=48_562
+INPUT_LEN=1675
+TARGET_LEN=373
```
Comment (thomasw21) [inline 4B8-en-ND-MLM.sh:24]: I'm thinking we need a way to compute these values from a given SEQ_LEN. Typically given a `noise_density`, `mean_noise_span_length`, and `sequence_length` we should be able to compute an input and target no? The reason why, is because what we really care about is that SEQ_LEN is 2048 (for performance), the rest we don';t really care as they are implementation details.

File: megatron/data/non_causal_mlm_dataset.py [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+        # T5-like span masked language modeling will fuse consecutively masked tokens to a single sentinel token.
+        # To ensure that the input length is `max_seq_length`, we need to increase the maximum length
+        # according to `masked_lm_prob` and `max_ngrams`. We can also define the label length accordingly.
+        expanded_inputs_length, targets_length = compute_input_and_target_lengths(
+            self.max_seq_length,
+            self.masked_lm_prob,
+            self.max_ngrams
+            )
+        self.expanded_inputs_length = expanded_inputs_length
+        self.targets_length = targets_length
```
Comment (thomasw21) [inline megatron/data/non_causal_mlm_dataset.py:179]: This should match your target_length that's specified in `--decoder-seq-length` no?

File: megatron/data/non_causal_mlm_dataset.py [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+        self.max_ngrams = max_ngrams
+        # T5-like span masked language modeling will fuse consecutively masked tokens to a single sentinel token.
+        # To ensure that the input length is `max_seq_length`, we need to increase the maximum length
+        # according to `masked_lm_prob` and `max_ngrams`. We can also define the label length accordingly.
+        expanded_inputs_length, targets_length = compute_input_and_target_lengths(
+            self.max_seq_length,
+            self.masked_lm_prob,
+            self.max_ngrams
+            )
+        self.expanded_inputs_length = expanded_inputs_length
```
Comment (thomasw21) [inline megatron/data/non_causal_mlm_dataset.py:178]: I think that's poor naming, that's the size of each row you need to load before converting it into MLM format. `number_of_raw_tokens` maybe?

File: train_ND_MLM_gpt.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    print_rank_0('> building train, validation, and test datasets for GPT ...')
+    # Option 1 of data loading using --data-path
+
+    if args.data_path:
+        train_ds, valid_ds, test_ds = build_train_valid_test_datasets(
+            data_prefix=args.data_path,
+            data_impl=args.data_impl,
+            splits_string=args.split,
+            train_valid_test_num_samples=train_val_test_num_samples,
+            max_seq_length=args.input_length,
```
Comment (thomasw21) [inline train_ND_MLM_gpt.py:200]: Yeah that's actually very confusing that you're passing input, instead of sequence_length ...

File: megatron/data/non_causal_mlm_dataset.py [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+        tokens_length += 1
+
+    inputs_length, targets_length = _tokens_length_to_inputs_length_targets_length(tokens_length)
+
+    # minor hack to get the targets length to be equal to inputs length
+    # which is more likely to have been set to a nice round number.
+    if noise_density == 0.5 and targets_length > inputs_length:
+        tokens_length -= 1
+        targets_length -= 1
+    return tokens_length, targets_length
```
Comment (thomasw21) [inline megatron/data/non_causal_mlm_dataset.py:433]:
```suggestion
    tokens_length = inputs_length

    while sum(_tokens_length_to_inputs_length_targets_length(tokens_length)) > inputs_length:
        tokens_length -= 1

    inputs_length, targets_length = tokens_length_to_inputs_length_targets_length(tokens_length)

    # minor hack to get the targets length to be equal to inputs length
    # which is more likely to have been set to a nice round number.
    if noise_density == 0.5 and targets_length > inputs_length:
        tokens_length -= 1
        targets_length -= 1
    # tokens_length is the number of raw tokens we need to get
    # inputs_length will be the input
    # targets_length will be the target
    return tokens_length, inputs_length, targets_length
```

So typically:
```python
>>> compute_input_and_target_lengths(2048, 0.15, 3)
(1860, 1675, 373)
```
And so you really need to only pass only the sequence length in `argument.py`

File: megatron/tokenizer/tokenizer.py [hunk 6]
```diff
@@ -348,3 +348,34 @@ def detokenize(self, token_ids):
     @property
     def eod(self):
         return self.tokenizer.eos_token_id
```
Comment (thomasw21) [inline megatron/tokenizer/tokenizer.py:350]: so we've overloaded `<eos>` as it signals:
 - end of document
 - seperation between input and target in case of MLM. 
 
 Since we already use eos as eod, can we create a new special token then? cc @SaulLu as she's our guru.

File: megatron/data/non_causal_mlm_dataset.py [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+        noise_density=0.15,
+        mean_noise_span_length=3
+        )])
+    labels_mask = ~mask_indices
+    
+    input_ids_sentinel = create_sentinel_ids(mask_indices.astype(np.int8), vocab_len=len(vocab_id_list))
+    labels_sentinel = create_sentinel_ids(labels_mask.astype(np.int8), vocab_len=len(vocab_id_list))
+
+    tokens = np.asarray([tokens])
+    input_tokens_ids = filter_input_ids(tokens, input_ids_sentinel, eos_id)[0]
```
Comment (thomasw21) [inline megatron/data/non_causal_mlm_dataset.py:243]: Typically this should use `eos_id` as we use it already to seperate documents.

File: megatron/data/non_causal_mlm_dataset.py [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+            for _start_idx, _end_idx in indices:
+                _len = _end_idx - _start_idx
+                if _len == max_len:
+                    samples_mapping.append([[doc_idx, _start_idx+_idx, _end_idx+_idx]])
+                else:
+                    sample_indices.append([doc_idx, _start_idx+_idx, _end_idx+_idx])
+                    current_len += _len
+
+        print_rank_0(' > done building sapmles index maping')
+        np.save(indexmap_filename, samples_mapping, allow_pickle=True)
```
Comment (thomasw21) [inline megatron/data/non_causal_mlm_dataset.py:328]: Unless I'm mistaken, you're missing a shuffling step here. I think it's quite important to shuffle it, otherwise you have a single document within a single batch and I believe @leandro and @loubnabnl have seen poor performances when you don't actually chuffle..

File: megatron/data/non_causal_mlm_dataset.py [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+            break_len = current_len + sample_len
+
+            indices = breakdown(sample_len, max_len=max_len)
+            for _start_idx, _end_idx in indices:
+                _len = _end_idx - _start_idx
+                if _len == max_len:
+                    samples_mapping.append([[doc_idx, _start_idx+_idx, _end_idx+_idx]])
+                else:
+                    sample_indices.append([doc_idx, _start_idx+_idx, _end_idx+_idx])
+                    current_len += _len
```
Comment (thomasw21) [inline megatron/data/non_causal_mlm_dataset.py:325]: Are you doing the same thing as `helpers.build_sample_idx`?

**Review by thomasw21 — COMMENTED** [review]
Had a quick look at MLMDataset. Thanks again for tackling this issue! Will read more when I have more time.

File: megatron/data/non_causal_mlm_dataset.py [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+    """
+    batch_size = input_ids.shape[0]
+
+    input_ids_full = np.where(sentinel_ids != 0, sentinel_ids, input_ids)
+    # input_ids tokens and sentinel tokens are >= 0, tokens < 0 are
+    # masked tokens coming after sentinel tokens and should be removed
+    input_ids = input_ids_full[input_ids_full >= 0].reshape((batch_size, -1))
+    input_ids = np.concatenate(
+        [input_ids, np.full((batch_size, 1), eos_id, dtype=np.int32)], axis=-1
+    )
```
Comment (thomasw21) [inline megatron/data/non_causal_mlm_dataset.py:387]: Wait this feels like `input_ids` would be 1676 instead of 1675. I don't see where you make sure that it's 1675 so that the final sequence length is 2048

File: megatron/arguments.py [hunk 11]
```diff
@@ -878,6 +878,8 @@ def __call__(self, parser, args, values, option_string=None):
                             'They are used for span masking in the T5 model')
     group.add_argument('--seq-length', type=int, default=None,
                        help='Maximum sequence length to process.')
+    group.add_argument('--input-length', type=int, default=None,
```
[the remark this answers is not in the document]
Comment (lintangsutawika) [reply megatron/arguments.py:881]: As far as I've understood, seq-length also initializes the max position embeddings. So for MLM, a seq-length of 512 would result in a final sequence length of 626. 

I didn't know how to handle this discrepancy so I opted to make a new `input-length` to handle 512 seq length and `seq-length` to handle the final length of 626.

File: megatron/data/non_causal_mlm_dataset.py [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+import torch
+
+from megatron import mpu, print_rank_0, get_tokenizer
+from megatron.data.blendable_dataset import BlendableDataset
+from megatron.data.dataset_utils import get_datasets_weights_and_num_samples, create_masked_lm_predictions
+from megatron.data.dataset_utils import get_train_valid_test_split_, get_split_by_range_, get_indexed_dataset_
+from megatron.data.indexed_dataset import make_dataset as make_indexed_dataset
+
+
+def build_train_valid_test_datasets(data_prefix, data_impl, splits_string,
```
[the remark this answers is not in the document]
Comment (lintangsutawika) [reply megatron/data/non_causal_mlm_dataset.py:33]: I adjusted `build_train_valid_test_datasets` input arguments.

And the `T5` dataset didn't count for prefix length. Also, my understanding is that `get_samples_mapping` in T5Dataset doesn't always return the required sequence length without paddings. So I adapted from the original T5 codebase and also made a `get_samples_mapping` that would concat every obtained sample to the desired input length.

File: prepare_tokenizer.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
+from transformers import AutoTokenizer, AddedToken
+
+tokenizer = AutoTokenizer.from_pretrained('bigscience/tokenizer')
+
+tokenizer.add_special_tokens({
```
[the remark this answers is not in the document]
Comment (lintangsutawika) [reply prepare_tokenizer.py:5]: I didn't want to changed that since each user might have different preference for their tokenizer.

File: 4B8-en-ND-MLM.sh (hunk 1 above)
Comment (lintangsutawika) [reply 4B8-en-ND-MLM.sh:24]: Yes, I agree. But I'm not sure where to put this function in.

File: megatron/data/non_causal_mlm_dataset.py (hunk 2 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mlm_dataset.py:179]: I was thinking of using `--decoder-seq-length` but I'm not sure if this would initialize other parts of the model?

File: megatron/data/non_causal_mlm_dataset.py (hunk 3 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mlm_dataset.py:178]: Thanks! I will use your suggestion.

File: train_ND_MLM_gpt.py (hunk 4 above)
Comment (lintangsutawika) [reply train_ND_MLM_gpt.py:200]: For `max_seq_length` here refers to the input prefix part and not the complete prefix+causal part. I guess I should changed the naming?

File: megatron/data/non_causal_mlm_dataset.py (hunk 5 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mlm_dataset.py:433]: Thanks, actually this solves the above arguments discussion. Thanks for the input.

File: megatron/data/non_causal_mlm_dataset.py (hunk 8 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mlm_dataset.py:328]: I was under the assumption that any pretraining data would already be shuffled or atleast not have any continuity between samples? Wouldn't shuffling during training also happen within the `dataloader` and not during data preparation in the dataset object?

I'm not sure how to shuffle in this part during initialization of the dataset.

File: megatron/data/non_causal_mlm_dataset.py (hunk 9 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mlm_dataset.py:325]: I'm concatenating sequences up the desired sequence length like that in `run_t5_mlm_flax.py`

https://github.com/huggingface/transformers/blob/6589e510fa4e6c442059de2fab84752535de9b23/examples/flax/language-modeling/run_t5_mlm_flax.py#L678-L691

I noticed that `get_samples_mapping` from `dataset_utils.py` doesn't always return a sequence with the desired length which then requires padding but that would mess the t5-style MLM.

File: megatron/data/non_causal_mlm_dataset.py (hunk 10 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mlm_dataset.py:387]: I can re-run to make sure.

File: megatron/data/non_causal_mlm_dataset.py (hunk 8 above)
Comment (thomasw21) [reply megatron/data/non_causal_mlm_dataset.py:328]: It's shuffled here for gpt dataset: https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/e23393fbefb86799de3960e03bb482361a504dbc/megatron/data/gpt_dataset.py#L376

So there's a few things:
 - document level shuffling, which can be assumed I guess
 - row level shuffling, ie once you create your segments of 2048 in length. Since you process documents sequentially, consecutive rows are often from the same document and that's usually an issue because you build a batch from consecutive rows. So in order to change that, gpt dataset had that shuffling index that allows you shuffle everything up together (should be the link I shared) before saving.

File: prepare_tokenizer.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (thomasw21) [reply prepare_tokenizer.py:5]: Hum but you are making some assumptions when building the dataset `self.sentinel_tokens = tokenizer.additional_special_tokens_ids` ...

File: megatron/data/non_causal_mlm_dataset.py (hunk 8 above)
Comment (thomasw21) [reply megatron/data/non_causal_mlm_dataset.py:328]: DataLoader should not be loading things in random order: https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/e23393fbefb86799de3960e03bb482361a504dbc/megatron/data/data_samplers.py#L93-L106

File: prepare_tokenizer.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (lintangsutawika) [reply prepare_tokenizer.py:5]: Oh yeah, which didn't get used. Maybe it's better to remove that line.

File: megatron/data/non_causal_mlm_dataset.py (hunk 8 above)
Comment (lintangsutawika) [reply megatron/data/non_causal_mlm_dataset.py:328]: Okay, I can add a shuffling process then.

File: megatron/data/non_causal_mlm_dataset.py (hunk 9 above)
Comment (thomasw21) [reply megatron/data/non_causal_mlm_dataset.py:325]: Hum this feels very close to what I expect `helpers.build_sample_idx` does. There's a python version of that same code: https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/e23393fbefb86799de3960e03bb482361a504dbc/megatron/data/gpt_dataset.py#L448

File: megatron/data/non_causal_mlm_dataset.py (hunk 8 above)
Comment (leandro) [reply megatron/data/non_causal_mlm_dataset.py:328]: @thomasw21 I guess you mentioned by here by accident.

Comment (thomasw21) [reply megatron/data/non_causal_mlm_dataset.py:328]: Oops @leandro my bad, I meant @lvwerra

Comment (lvwerra) [reply megatron/data/non_causal_mlm_dataset.py:328]: Yes, we saw that very long documents in the pretraining corpus could dominate a single batch leading to significantly slower convergence.

File: megatron/data/non_causal_mlm_dataset.py [hunk 14]
```diff
@@ -0,0 +1,499 @@
...
+            breakdown(sample_len, idx_offset=idx_offset, idx_list=idx_list)
+
+        idx_list = [0]+idx_list
+        return list(zip(idx_list[:-1], idx_list[1:]))
+
+
+    # Filename of the index mapping
+    indexmap_filename = data_prefix
+    indexmap_filename += '_{}_indexmap'.format(name)
+    indexmap_filename += '.npy'
```
Comment (thomasw21) [inline megatron/data/non_causal_mlm_dataset.py:282]: Can we change the naming, something that mentions it's mlm dataset? Ideally we could have a index for gpt dataset and one for mlm dataset based on the same dataset.

Comment (thomasw21) [reply megatron/data/non_causal_mlm_dataset.py:282]: Also stupid question, why do you have only one file, GPT-dataset seems to generate 3 files: samples / index / shuffle.

File: prepare_tokenizer.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (thomasw21) [reply prepare_tokenizer.py:5]: They did get used (otherwise your MLM implementation sounds wrong, if you don't provide a list of sentinel tokens)

File: megatron/data/non_causal_mlm_dataset.py [hunk 15]
```diff
@@ -0,0 +1,499 @@
...
+        return len(self.samples_mapping)
+
+    def __getitem__(self, idx):
+
+        indices = self.samples_mapping[idx]
+        sample = []
+        for doc_idx, start_index, end_index in indices:
+            sample.append(self.indexed_dataset.get(doc_idx)[start_index:end_index])
+
+        return build_training_sample(
```
Comment (thomasw21) [inline megatron/data/non_causal_mlm_dataset.py:212]: I think we should cache that result. ie create a `number_of_samples x 2048` array with already preprocessed MLM. I can clearly see that step being a bit slow, and so loading directly from an array should make things much faster?

**Comment by Muennighoff** [comment]
This PR has been fully superseded by [mlm](https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/287) & [mtf](https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/288), correct?

**Comment by lintangsutawika** [comment]
Yes, we can close this if needed.

**Comment by thomasw21** [comment]
Let's keep it. There's still the code from train scripts.
